### ISSUE TITLE
feat: add database integration with sqlalchemy

### ISSUE THREAD

**Issue by ajndkr** [ajndkr/lanarky#84]
New milestone feature: db integration with sqlalchemy

Useful for storing user conversations, tracking token usage, and so on.

### Acceptance Criteria

- [ ] add lifespan event for sqlalchemy session
- [ ] add basic schemas and models

**Comment by awtkns** [comment]
@ajndkr would you be able to elaborate on this? Is lanarky meant to solely focused on enabling streaming or does it have a broader goal?

**Comment by ajndkr** [comment]
@awtkns yeah sure. So the project is still holds streaming as the key features but from my experience, there are other common functionalities which can be used via lanarky. The broader goal is to provide necessary tooling for deploying without having to build them from scratch.

For example, tracking tokens per user question is something which is not trivial to build but it needs an sql connection. Hence this particular issues.

**Comment by npuichigo** [comment]
user id, conversation id or session id may also be useful for vector store retrieving

**Comment by ajndkr** [comment]
hi! this feature is still part of the roadmap but i'm unable to find time currently. would you like to contribute?

**Comment by npuichigo** [comment]
@ajndkr glad to contribute if have time.

One question, currently the langchain object is created before passing to the `LangChainRouter` and wrapped in a simple `create_langchain_dependency` for dependency injection. Actually, langchain already has session support for memory using `sqlalchemy`, `redis` or `motorhead`, but in order to bind with some user session id, the dependency injection should be more flexible.
For example, instead of the simple `create_langchain_dependency`, maybe it's better for user to pass the dependency, which may use the `session_id` in input query or body to construct the memory used for langchain object. Do you have any insights on the ideal API for such things?

**Comment by npuichigo** [comment]
@ajndkr In current implementation, the request and response model are dynamically determined by `langchain_object` with `create_request_from_langchain_dependency`. What if the constrcution `langchain_object` needs to parse request first to get something like `session_id` for db memory? Can u give me some advice on this?

**Comment by ajndkr** [comment]
@npuichigo hi! `create_request_from_langchain_dependency` as a feature isn't fully mature to handle all use cases. We add new parameters to handle new requirements but in the meantime, you can check this implementation: https://github.com/menloparklab/chatbot-ui/blob/d2d4aa84ebb6351bfacee99429228d091221b60b/backend/app.py#L101-L114

You can directly use the StreamingResponse class and add custom logic in the wrapped function.

**Comment by npuichigo** [comment]
The code linked uses all the features except `add_langchain_api_route` part of `LangchainRouter`. So is that also immature since it uses `create_request_from_langchain_dependency`?

**Comment by ajndkr** [comment]
> The code linked uses all the features except `add_langchain_api_route` part of `LangchainRouter`. So is that also immature since it uses `create_request_from_langchain_dependency`?

nope! the linked code is more low level, so you can skip the use of `add_langchain_api_route` and use the base fastapi route decorator.
